packages/hurl/src/report/tap/mod.rs

```diff
@@ -0,0 +1,234 @@
+/*
+ * Hurl (https://hurl.dev)
+ * Copyright (C) 2023 Orange
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *          http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+use crate::report::Error;
+use crate::runner::HurlResult;
+
+use std::fs::File;
+use std::io::Write;
+
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct Testcase {
+    description: String,
+    success: bool,
+}
+
+impl Testcase {
+    /// Creates an Tap &lt;testcase&gt; from an [`HurlResult`].
+    pub fn from(hurl_result: &HurlResult, filename: &str) -> Testcase {
+        let description = filename.to_string();
+        let success = hurl_result.errors().is_empty();
+        Testcase {
+            description,
+            success,
+        }
+    }
+
+    /// Creates an Tap &lt;testcase&gt; from a TAP line
+    /// ok 1 - this is the first test
+    /// nok 2 - this is the second test
+    fn parse(line: &str) -> Result<Testcase, Error> {
+        let mut line = line;
+        let success = if line.starts_with("ok") {
+            line = &line[2..];
+            true
+        } else if line.starts_with("nok") {
+            line = &line[3..];
+            false
+        } else {
+            return Err(Error {
+                message: format!("Invalid TAP line <{line}> - must start with ok or nok"),
+            });
+        };
+
+        let description = match line.find('-') {
+            None => {
+                return Err(Error {
+                    message: format!("Invalid TAP line <{line}> - missing '-' separator"),
+                })
+            }
+            Some(index) => {
+                if line.split_at(index).0.trim().parse::<usize>().is_err() {
+                    return Err(Error {
+                        message: format!("Invalid TAP line <{line}> - missing test number"),
+                    });
+                }
+                line.split_at(index).1[1..].trim().to_string()
+            }
+        };
+        Ok(Testcase {
+            description,
+            success,
+        })
+    }
+}
+
+/// Creates/Append a Tap report from a list of `testcases`
+pub fn write_report(filename: &str, new_testcases: &[Testcase]) -> Result<(), Error> {
+    eprintln!("write tap report {filename}");
+    let mut testcases = vec![];
+
+    let existing_testcases = parse_tap_file(filename)?;
+    for testcase in existing_testcases {
```

Comment:
Cloning testcase can be avoided:

```rust
for testcase in existing_testcases.iter() {
    testcases.push(testcase);
}
for testcase in new_testcases.iter() {
    testcases.push(testcase);
}
```
And the signature of `write_tap_file`should be change accordingly:
`pub fn write_tap_file(filename: &str, testcases: &[Testcase]) `to `pub fn write_tap_file(filename: &str, testcases: &[&Testcase]) `

Reply:
I'm not sure about changing  the signature to `&[&Testcase]`
currently we already use `&[Testcase]` for the `write_report` function for the 3 types: html, junit and tap.

Reply:
`write_tap_file` should be private though